packages/belos/tpetra/test/MINRES/test_minres_diag.cpp

```diff
@@ -0,0 +1,435 @@
+
+  RCP<const Teuchos::Comm<int>> comm = Tpetra::getDefaultComm();
+  const int pid = rank(*comm);
+
+  bool verbose = false;
+  bool success = true;
+
+  try {
+
+    int n(10);
+    int numRHS=1;
+    
+    RCP<MP> map = RCP(new MP(n, 0, comm));
+    
+    MV X(map, numRHS), Y(map, numRHS);
+    X.putScalar( 1.0 );
+    
+    // Inner computes inv(D2)*y
+    RCP<Diagonal_Operator_2<MV>> D2 = rcp(new Diagonal_Operator_2<MV>(n, map->getMinGlobalIndex(), 1.0));
+    Iterative_Inverse_Operator<OP, ST, MP, MV> A2(n, 1, D2, "Belos (inv(D2))", true);
+    
+    // should return x=(1, 1/2, 1/3, ..., 1/10)
+    // AM: If you uncomment this line, test failed, I don't know why
+    // A2(X,Y);
+    
+    if (pid == 0) {
+      std::cout << "Vector Y should have all entries [1, 1/2, 1/3, ..., 1/10]" << std::endl;
+    }
+    Y.print(std::cout);
+
+    // Inner computes inv(D)*x
+    RCP<Diagonal_Operator<MV>> D = rcp(new Diagonal_Operator<MV>(n, 4.0));
+    RCP<Iterative_Inverse_Operator<OP, ST, MP, MV>> Inner =
+      rcp(new Iterative_Inverse_Operator<OP, ST, MP, MV>(n, 1, D, "Belos (inv(D))", false));
+
+    // Composed_Operator computed inv(D)*B*x
+    RCP<Diagonal_Operator<MV>> B = rcp(new Diagonal_Operator<MV>(n, 4.0));
+    RCP<Composed_Operator<MV>> C = rcp(new Composed_Operator<MV>(n, Inner, B));
+
+    // Outer computes inv(C) = inv(inv(D)*B)*x = inv(B)*D*x = x
+    RCP<Iterative_Inverse_Operator<OP, ST, MP, MV>> Outer =
+      rcp(new Iterative_Inverse_Operator<OP, ST, MP, MV>(n, 1, C, "Belos (inv(C)=inv(inv(D)*B))", true));
+
+    // should return x=1/4
+    (*Inner)(X,Y);
+
+    if (pid == 0) {
+      std::cout << std::endl << "Vector Y should have all entries [1/4, 1/4, 1/4, ..., 1/4]" << std::endl;
+    }
+    Y.print(std::cout);
+
+    // should return x=1
+    (*Outer)(X,Y);
+
+    if (pid == 0) {
+      std::cout << "Vector Y should have all entries [1, 1, 1, ..., 1]" << std::endl;
+    }
+    Y.print(std::cout);
+
+    // Compute the norm of Y - 1.0
+    std::vector<ST> norm_Y(Y.getNumVectors());
+    Teuchos::ArrayView<ST> normView(norm_Y);
+
+    Y.update(-1.0, X, 1.0);
+    Y.norm2(norm_Y);
+
+    if (pid == 0)
+      std::cout << "Two-norm of std::vector (Y-1.0) : "<< norm_Y[0] << std::endl;
+
+    success = (norm_Y[0] < 1e-10 && !Teuchos::ScalarTraits<ST>::isnaninf( norm_Y[0] ) );
+    
+    if (success) {
+      if (pid == 0)
+        std::cout << "End Result: TEST PASSED" << std::endl;
+    } else {
+      if (pid == 0)
+        std::cout << "End Result: TEST FAILED" << std::endl;
+    }
+  }
+  TEUCHOS_STANDARD_CATCH_STATEMENTS(verbose, std::cerr, success);
+
+  return (success ? EXIT_SUCCESS : EXIT_FAILURE);
+}
+
+//************************************************************************************************
+//************************************************************************************************
+
+int main(int argc, char *argv[])
+{
+  run<double>(argc, argv);
+  // run<float>(argc, argv); // FAILS
```

Comment:
I appreciate the effort to make tests with multiple scalar types! This will be needed in the future but is not currently a capability we rely on. 
That said, this test needs 
1. to be wrapped with a check to make sure float is enabled with the build (CMake option Trilinos_ENABLE_FLOAT=ON) and
2. to adjust the convergence tolerance based upon the scalar type. A solver run in float will not converge to 1e-10 [see line 413] as the solver in double will. In my experience, 1e-4 is sufficient tolerance for a single precision solver.  Or you could set the tolerance to something like sqrt(machine epsilon).